--- 
title: "Quantitative skills for biology"
# author: "Yolanda Wiersma, Amy Hurford, Joany Mari√±o"
# date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook #bookdown::html_document2
documentclass: book
bibliography: [references.bib]
biblio-style: apalike
link-citations: false
github-repo: ahurford/quantitative-training-resources/tree/master/Quantitative training guides/quant-guide
description: "Quantitative skills for biology"
---

# Introduction {#intro}

Quantitative skills are necessary in biology. Quantitative skills include performing statistical analyses, mathematical modelling, managing and reformatting data, designing experiments, interpreting graphs, vizualizing data, identifying genes from DNA sequences, constructing 3D models of proteins, and making phylogenies. All of these tasks are now performed using computer software, and elements of computer programming, statistics, and applied mathematics can be useful for professional biologists. Quantitative skills and best practices in data science are now being incorporated throughout the biology major.

Teaching quantitative skills is necessary since a survey reported that more than 90% of ecologists found, that in retrospect, their training in statistics and mathematics, during their undergraduate biology major, was too low [@barraquand2014lack]. This same study found that practicing ecologists recommended that 30% of coursework during a biology major be dedicated to quantitative training.

Today's biologists use a variety of softwares, and particularly, `Microsoft Office`, `R`, `Python`, and `ArcGIS`. Much of this quantitative training manual teaches data manipulation and visualization in `R`. Our choice of `R` is for the following reasons:

1. A valued skill: As of March 2019, `R` is the [fifth ranked software](http://r4stats.com/articles/popularity/) listed in job advertisements for Data Scientists. `Python` is the number one ranked software, but `R` is a popular statistical software and is popular in biology due to the need to analyze the results of experiments using statistics.

1. Reproducibility: `R` is a scripting software, so unlike _point and click_ softwares, the steps to produce a data analysis or figure are fully reproducible.

1. Accessibility: `R` is free.

1. No limits: `R` has a large user community and contributed packages allow you to do almost anything. `R` packages range from [serious](https://rviews.rstudio.com/2017/10/09/population-modeling-in-r/) to [fun](https://www.r-bloggers.com/useless-but-fun-r-packages/)).

1. Popularity: A survey of studies from of 30 ecology journals found that in 2017, 58% of article used `R` as the primary tool to generate their results [@lai2019evaluating].

<!--chapter:end:index.Rmd-->

# How to install R and R Studio {#install}

`R` can be installed from https://www.r-project.org/. If you have problems with installing `R` this same website provides advice on how to seek support.

`R` is a scripting language. `R Studio` is a graphical user interface (GUI) that facilitates `R` coding by providing buttons and menus to provide options for some commonly used commands. Download the free version `RStudio Desktop` from https://rstudio.com/products/rstudio/

For further help installing `R` and `RStudio` see [YaRrr! The Pirate's Guide to R](https://bookdown.org/ndphillips/YaRrr/installing-base-r-and-rstudio.html)

<!--chapter:end:01-installing-R.Rmd-->

# Finding your way around RStudio {#rstudio}

R studio has 4 windows (also known as "panes") that are inter-linked. The default layout is shown in Figure \@ref(fig:panes). You can move the layout of the panes and change their size and shape by clicking and dragging with your mouse along the window pane borders. The details of what happens in each pane are detailed below.

```{r panes, fig.cap="Panes in RStudio. 1. Source pane, to work with R scripts (code won't be evaluated until you send it to the console). 2. Console pane, where code is executed by R. 3. Environment/History pane, to view objects in the working space and command history, respectively. 4. Files/Plots/Packages/Help pane, to see the file directories, plots, installed packages and access R help, respectively.", fig.align='center', out.width="70%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/Fig_5_1.jpg')
```

## Console pane

The Console pane (labelled 2 in Figure \@ref(fig:panes)) is where code is evaluated (also referred to as executed). The Console pane has a prompt (the `>` symbol), which tell you that R is ready to receive input. You can type code here to evaluate it and get an output.

**TRY IT!** Type 5 + 5 in the console at the prompt and press `enter` or `return`. You'll see that R gives the correct output (Don't worry about the [1] for now, we'll discuss what this means in Chapter X).

```{r}
5+5
```

Use the console to test and query code, or do a quick analysis that you do not want to save. Press the up arrow to scroll through past commands, from most to least recent.

To do more interesting things, we need to assign values to objects. For this, we create an object by specifying a name followed by the assignment operator `<-` and the value we want to give it. For example:

```{r}
x <- 5
```

When you run this code by pressing `enter/return` you will not see an output as you did before. But, you can ask R for the value by typing the object name and pressing `enter/return`:

```{r}
x 
```

Now that we have saved the value of `x`, we can do arithmetic operations with it. For example, instead of adding two numbers, we can assign those values to objects and then add them.

**TRY IT!** Add 3 and 8 by assigning the values to objects. First assign 3 to `x` and 8 to `y`. Then add `x+y`.

You can overwrite the value of objects The current value of your object is the last value you gave it. In the console type:

```{r}
x <- 2
```
Press `enter/return`. Now type:

```{r}
x <- 1
```

Press `enter/return`. What is the value of

```{r}
x + 1
```

Type:

```{r}
x
```

to query the current value of `x`.

----

EXCERCISE

Can you write the code, and execute the commands in the correct order so that `x + 1 = 6`?

----

## Source pane

The Source pane (labelled 1 in Figure \@ref(fig:panes)) is where you create, save, and edit `R` scripts. `R` scripts are blocks of code that perform a task that you want to save (e.g.., are the answer to the EXERCISE above, or import and clean data to make a figure) and when saved have a `.R` extension. When you've written and saved a script, you can re-use it for future projects. 
Try it now: open and save a new `.R` file to save your work in. In the top menu select `File > New File > R Script`, then `File > Save As`. Paste the following `R` code into your script and save it:

```
x <- 4
x <- 6
```

Now, you have saved this code, but has it been evaluated? To answer this question type `x` in your `Console` and press ENTER:

```{r}
x
```

It is important to note that saving commands in a script in your `Source` window does not evaluate the commands in your `Console`. To execute the commands, with your script displayed in the `Source` pane, you need to click the `Source` button (righthand side of the source pane, equivalently select `Code > Source` from the top menu bar; note, if you have a small `Source` pane the `Source` button may not appear and you may need to make it bigger or use the menu bar).

`Source` your script and then evalate `x` in your `Console`. If you have done this task correctly the value of `x` should be `6` and it will appear as `[1] 6` in your `Console`.

What happened to `x <- 2`? When `Source` is called the commands are executed in the `Console` from top to bottom as they appear in the script. Therefore, `x <- 2` was only the current value of `x` for a very short time before it was overwritten with the command `x <- 6`. 

---

EXERCISE

Edit your script so that now it reads:

```
x <- 4
y <- x + 1
x <- 6
z <- x + 1
```

`Source` the script and then query the values of `y` and `z` in your `Console`.

```
> y
> z
```

Explain why the values of `y` and `z` are `5` and `7` respectively.

---

If you would like to run one line or several lines of your script only (not all the lines), then place your cursor on the line you want to execute and click the `Run` button (equivalenty `Code > Run Selected Line(s)`, or press `Ctrl`+`enter` or `Cmmd`+`return` for a shortcut).

## Enviornment/History pane

This pane has several tabs, we'll discuss each in turn. 

### Environment tab {-}

This shows you the names of all data objects (e.g., vectors, matrices, dataframes) that you've defined in the current R session, along with information about the size of data objects. 

There are also clickable actions, like `Import data set`, which we'll refer to in Chapter \@ref(data-entry). 

### History tab {-}

This shows you a history of all the code you've previously run in the console. From here you can 
re-run this commands by selecting them and clicking `To Console` or include them in your script by clicking `To Source`.

**TRY IT!** Select one of the commands and click `To Console`. You'll see the command is executed and the output is given. Now, try to include another command in the script that you have been working with during this session. 

## Files/Plots/Packages/Help/Viewer pane

This pane also has a few tabs. We'll go through each one in detail.

### Files tab {-}

This tab gives you access to the file directory on your own computer. It's a useful spot to set your working directory. For this, you have to navigate to the folder whence you want to read and save files, then click `More` and select `Set As Working Directory` (see Chapter \@ref(data-entry) for more details). 

**TRY IT!** Navigate through the file tree and find the folder that you want to work with, alternatively, create a new one by clicking `New Folder`. Now, open the folder and make it your working directory. You'll see that in the console R runs the command for changing the working directory.

### Plots tab {-}

This tab lets you see all the plots (graphs) that you create. You can also uset the buttons at the top to zoom, and to export the graph as a PDF or jpeg file. If you make multiple plots, you can navigate through them by clicking on the left and right arrows. We will come back to this tab in Chapter \@ref(graph).

### Packages tab {-}

This tab shows a list of all the R packages installed on your computer (see Chapter X for more on packages). It also indicates which packages are loaded with a check mark. You can browse through the installed packages by typing the name in the blank search field.

Packages are add-on functions and datasets that you can add to your `R` installation. To install a package you can click on `Install`, this feature will open a window with a blank field where you can type the name of the package. When you are ready click on `Install` to proceed with the installation. Alternatively, you can install with the command `install.packages("name")`, where ‚Äúname‚Äù is the name of the package. When you install a packege in either way, you will see in the console that R displays some information about where the package is being downloaded from and a progress bar. 

**TRY IT!** One package that we will use later on is called `ggplot2`, install it by either using the `install.packages` command or the `Install` feature. 

### Help tab {-}

Here you can get help with R functions. You can type the name of the function in the search field (blank bar with the magnifying glass beside it) to see the related help page. Equivalently, you can use code to perform the search by typing `?` followed by the function name. 
See Chapter \@ref(help) "Getting R help" for more details on reading the help files.

**TRY IT!** Find the help page for the `plot` function, either by typing the function name in the search field or directly into the console. We will use this function in Chapter \@ref(graph).

### Viewer tab {-}

This tab is similar to the Plots tab, but it is used to view local web content. For example, to display generated web content or applications that you are developing. This is a more advanced feature and we will not cover it in this guide.

Add links to other resources

<!--chapter:end:02-finding-your-way.Rmd-->

# Getting R help {#help}

The first place most people go for R help is the "Help" tab in the "Files/Plots/Packages/Help" pane on the console (described in Chapter \@ref(rstudio)). Here you can do a search by topic/term using the search window at the top right. You can also click the "house" icon to go to the home menu for the help files and use the table of contents there to navigate to what you are looking for.

## Using R help

The hardest part about R help files is that at first they don't seem that helpful, and to new R users, they seem to be written in code. However, all of the help pages are structured the same way, so once you learn how to de-code them, you'll be well on your way. Below is an example of how to interpret R help files

## R "Cheatsheets"

Because R is open source software, there is also a culture of open access help. Several R programmers have made helpful "cheatSheets" that are worth printing off and hanging by your desk. Try [this one](https://cran.r-project.org/doc/contrib/Short-refcard.pdf) by Tom Short for starters.

## R books

As with the cheatsheets, there are open access books out there (as well as expensive ones to buy). For beginners, "The Pirates Gude to R" by Nathaniel D. Phillips is **HIGHLY** recommended. You can access it [here](https://bookdown.org/ndphillips/YaRrr/) 

## How to interpret R help files

R documentation follows the following format. As an example, we'll use the screenshot below for a really simple statistical command (the t-test) to walk you through the R documentation. Numbers below correspond to the labels on the screenshot.

1. **header** This simply gives the command for the thing you are trying to do, in R language

2. **Title** This is a title of the function/command/tool in more plain language

3. **Description** This section provides a short description of what the command does, written in reaonably plain language

4. **Usage** This is where you see an example of R code that you use to execute the command (**TIP** you can tell what text in the help file is R code and what is other information by the font. Example code is in Courier font). Any example prefaced with two hashtags ("##") is a comment that adds additonal description or provides different options for the comment. For example, for the t-test here, the first exampe is for the default method, where you are doing a t-test for a difference between two numeric vectors of data values ("x" and "y"). The second example is where you are applying the t-test to a formula instead of two numeric vectors. Until you start doing more advanced statistics, the default method will usually be what you use.

5. **Arguments** This section breaks down every element in the R code under "Usage" above it and describes what the input is that you need to provide. It also gives you details about the options as well as what the defaults are. For example, for the t-test documentation here, it tells you that "x" is "a (non-empty) numeric vector of data values". It also tells you that this is what "y" is but note that it tells you that "y" is optional (i.e., you can do a t-test on just one vector of data values). This fact is also shown in the example code where it says "y = NULL").

Note that you often do not have to input ALL the arguments. If you are happy with the defaults you don't have to type in the code. For example, the description of the argument "alternative" tells you that you can specify whether the alternative hypothesis is two-sided (i.e., two-tailed), or one-tailed, in which case you have to say whether the alternative hypothesis is that the second set of data (e.g., the treatment) are greater or less than the first (e.g., the control). The default is that you are assuming the alternative hypothesis is two-sided. If this is what you want, then your code only needs to look like this

``` t.test(x, y)```

However, if you want a one-sided test, and your alternate hypothesis is that the treatment (y) is greater than the control (x), then you need to type a bit more code, like this: 

```t.test(x, y, alternative = c("greater"))``` 


6. **Details** This sectin gives you a little more detail (written a bit more like a stats text) about what the function is doing. For more advanced functions, there may be details on the types of inputs allowed and other caveats. *This section is really helpful to read closely!*

7. **Value** This section lists the objects (output) created when you execute the command. The relevance of this list will become more clear once you start using R.

8. **See Also** This section is usually just links to similar commands to the one you are reading up on, or related to it somehow. These can be useful to explore when doing more complex statistical tests, or if you are not sure which test you should be using. Clicking the link will just take you to the Documentation page for that command, which will be set up exactly the same way as this one.

9 **Examples** This gives sample code and output. If you have installed the necessary package(s) to run this code, you can just try the code out yourself by typing the code *exactly* as shown in the example in the Console window of RStudio. It will call data that comes built in with the package when you installed it and can be a helpful way for you to understand how the command is working before you try it with your own data. 


```{r fig.align='center', echo=FALSE, message=FALSE, out.width="95%"}
knitr::include_graphics("./figures/HelpDoc_partA.jpg")
```

```{r fig.align='center', echo=FALSE, message=FALSE, out.width="95%"}
knitr::include_graphics("./figures/HelpDoc_partB.jpg")
```

## Debugging

- Spelling mistakes
- Input to a function is the wrong data type
- Run line by line
- Cut and paste example


<!--chapter:end:03-getting-R-help.Rmd-->

# Data and metadata {#data}

Joany will write this section. Maybe we could do a lab meeting or eco-evo to figure out what to do for metadata, i.e., look for good examples.

I think look at the Ethan White course. The paper by Kara Woo is quite good and worth a cite.

Here we need to provide the rationale for saving data as we recommend we also need to provide the metadata guidelines. Maybe find a paper by someone who is up on this kind of stuff and see how they did the metadata. Also talk about why highlight cells in excel or making borders isn't a good idea, i.e. tidy data exercise.

Twitter recommends:
https://github.com/ropenscilabs/dataspice

https://aslopubs.onlinelibrary.wiley.com/hub/journal/23782242/about/author-guidelines#data

https://www.usgs.gov/products/data-and-tools/data-management/metadata-creation

Browse http://sciencebase.gov and http://data.usgs.gov for both good and bad examples fo these standards in use (as well as examples from before)  http://Data.gov will have more examples too

The metadata sectin of this paper is really great for a quick run down of what metadata is: https://ojs.library.queensu.ca/index.php/IEE/article/view/4608
 

<!--chapter:end:04-data-and-metadata.Rmd-->

# Introduction to R {#rintro}

This is a brief introduction to R, focussing on basic programming concepts and best practices.

## Workflow

You can interact directly with the console, but it is considered best practice to write an .R file and run commands in the console from that script. This allows saving the code and running all at once with the `source` function. Remember to click Save often, so that you actually save your script!


Once you open an R script (or any file), an editor panel will appear in the top left. From there you can run the current line (or a selected block of code) by clicking on the `Run` button in the editor panel, or in the `Code` menu select `Run Lines`. There is also a `Re-run` button next to the `Run` button, which allows to run again the previous code block including any new modifications you have just made.

## Working directory

The working directory is a folder on your computer where R will look for data, save your plots, etc. To make your workflow easier, it is good practice to save everything related to one project in the same place, as it will save you time typing up computer paths or finding the files across your computer. For example, you could save your script and all the data for this exercises in a folder called "Intro-to-R". It is good practice to avoid spaces in file names because it can confuse R.

To find out where your working directory is now, run the function `getwd()`. If you want to change it, you can use `setwd("dir")`, where `"dir"` is the path to the folder that you want to use. Notice that the path must be written inside quotation marks. You'll see that in the console R runs the command for changing the working directory. 

Alternatively, you can set the working directory from the Files tab in the Files/Plot/Packages/Help pane. For this, navigate to the folder whence you want to read and save files or create a new one by clicking `New Folder`. Once you have located the folder, click on it to access it. Then, click `More` and select `Set As Working Directory`. 

**TRY IT!** Set the working directory to a folder in your Desktop called "Intro-to-R". 

## Variables and assignment

The assignment operator is `<-`. It stores whatever you write on the right to variables on
the left. Variable names can be made up of letters, numbers, underscores and periods. But, they cannot start with a number nor contain spaces. For instance:

```{r}
x <- 10
```

Notice that assigning a variable does not print its value, 
but if you look at the `Environment` panel, you will see the new variable there.
You can also check that it is stored by calling the variable:

```{r}
x 
```

**TRY IT!** Define a varible "y" with the value of 3. 

### Shortcuts

In RStudio, the assignment operator is written in a keystroke by hitting `Alt`+`-`.

A practical way of running code from a script by using a keyboard shortcut. In GNU/Linux and Windows, you can hit `Ctrl`+`Return`, on macOS the equivalent is `‚åò Command`+`Return`. This shortcut also allows running a selected block of code. 

You can alternate between the script and console panels with the `Ctrl`+`1` and `Ctrl`+`2` shortcuts.

From the console, you can go back to the previous command by hitting the `up` arrow. If you press the `up` arrow multiple times you will walk back through the history. You can navigate the history by pressing the `down` arrow as well. You can edit any command from there by using the `left` arrow or `delete` keys.

## Using variables

Variables can be used in calculations

```{r}
x + x
```

Variables can also be reassigned

```{r}
x <- 314
x
```

## Comments

In R, the `#` sign is a special character that indicates a comment. Anything that is written directly to the right of a `#` sign will be ignored by R. Comment your scripts often!

**TRY IT!** In your script, write a couple of sentences to explain each of the commands we have used so far.

## Precedence

The order of operations follows the standard precedence (from highest to lowest):

* Parentheses `(`, `)`
* Exponents or roots: `^`
* Multiplication or division: `*`, `/`
* Addition or subtraction: `+`, `-`

**TRY IT!** Using R, can you calculate if the result of `1 + (2 * 3 * 5)` is the same as `(1 + 2) * 3 * 5`? 

## Mathematical functions

R comes with many built in mathematical functions. Functions are called by typing its name followed by open and closing parentheses. Whatever we write inside the parentheses are the function's arguments:

```{r}
sqrt(x)
```

**TRY IT!** Calculate the natural logarithm of `x`. Hint: the function to do so is `log()`, and the argument would be the name of the variable that we want to use.

## Data types

There are 6 basic data types in R:

* Character: `h` or `hello`.
* Numeric (real or decimal): `3` or `315.5`.
* Integer: `4L` (the letter L specifies that this is an integer).
* Logical: `TRUE` or `FALSE`.
* Complex: `7+4i`.

## Data structures

Elements of these data types may be combined to form data structures. We will look at three of them.

### Vectors {-}

A *vector* is a set of values, in a specific order, and of the same data type. To create a vector we can use the function `c()`, where the "c" stands for concatenate or combine.
For example:

```{r}
rgb <- c("red", "green", "blue")
rgb
y <- c(2, 3, 5, 7)
y
```

To check what kind of data object it is you can use the funcion `class`. To see check how long is the object you can use `length()` for one-dimensional objects, or `dim()` for objects of two or more dimensions. For example:

```{r}
class(rgb)
length(rgb)
```

**TRY IT!** What is the data type and length of the vector `y`?

The core idea behind vectorization is that operations can be applied at once 
to an entire set of values. For example:

```{r}
y * 3
```

**TRY IT!** Create a vector with a sequence of numbers using the function `seq`. Hint: run `?seq` in the console to display the help page of the function.

A key point of the previous example is that vectors can be used as arguments to functions. But, vectors are *atomic*, which means that they can only contain one class of data. In other words, different classes cannot be mixed in the same object. For instance, if we combine numeric with character data types, R will coerce all elements to be characters:

```{r}
primes <- c(2, 3, 5, "seven")
class(primes)
```

For this reason, and to avoid errors, it is important to check frequently the data type of the objects that you are working with.

### Matrices {-}

In R, matrices are a type of vectors with two dimensions: the number of rows and columns. As with vectors, the elements of a matrix must be of the same data type (numeric or character). One way of creating a matrix with the function `matrix()`. Keep in mind that matrices are filled column-wise:

```{r}
M <- matrix(
  c(4, 5, 6, 7),
  nrow = 2,
  ncol = 2)
M
```

We can confirm the type of data with the function `class()`, and look at the size of the matrix with the function `dim()`:
```{r}
class(M)
dim(M)
```

**TRY IT!** Create a 3 by 3 matrix in which the elements are the numbers from zero to nine.

### Lists {-}

A list is a special type of vector, where each element can be of a different data type. To create a list we can use the function `list()`:

```{r}
my_list <- list(5, "hello", TRUE, 1+4i)
my_list
class(my_list)
```

As you can see from the output, each element of the list starts on a new line.
Lists can be useful inside functions because you can combine different kinds of results into a single object that a function can return.

**TRY IT!** What is the lenght of the object `my_list`?

### Data frames {-}

A data frame is a type of list in which every element of the list has the same length, that means that a data frame is a ‚Äúrectangular‚Äù list. We can create a data frame with the function `data.frame()`:

```{r}
df <- data.frame(id = letters[1:10], x = 1:10, y = 11:20)
df
```

To get some information on our data frame, we can use again the function `dim()`. Other useful functions for data frames are `nrow()`, `ncol()`, `head()`, and `tail()`.

**TRY IT!** Create a dataframe as especified above. What information you can get from the functions `nrow()`, `ncol()`, `head()`, and `tail()`.

## R Packages

Functions can be added to R with packages, which are stored on [CRAN](https://cran.r-project.org/web/packages/index.html) (the Comprehensive R Archive Network). 

To install a package you use the function `install.packages("packagename")`, where `packagename` is the name of the package you want to install. To see what packages are installed by type `installed.packages()`. Packages are loaded into the R session with the `library(packagename)` function. 

As an alternative, you can install a package through the Packages tab in the the Files/Plots/Packages/Help pane. For this, click on `Install`, this feature will open a window with a blank field where you can type the name of the package. When you are ready click on `Install` to proceed with the installation. You will see in the console that R displays some information about where the package is being downloaded from and a progress bar. You can also load a package from this tab. For this, find the the package that you want to load either by scrolling trough the list or by writing its name on the search bar to the right of the tab. Once you find it, click on the checkbox to the left of the package name. You will see that R runs the command to load the package in the console. 

**TRY IT!** Install the package `ggplot2`, by either using the `install.packages` command or the `Install` feature. Then, load the package either with the `library()` or the checkbox. 


<!--chapter:end:05-R-intro.Rmd-->

# Entering and loading data {#data-entry}

## Entering data

Good data organization is the foundation of any research project, 
and this begins with entering and archiving data. 
There are several reasonable options for data entry, for example:

1. Spreadsheet
2. Text file
3. Database
4. Form (web or GUI databases). 

In R, you can also input data directly, which can be the right choice if you are working with a small dataset. 
Here, you will learn how to enter the data from your experiments
and save them using R. 

Let's start by
making Table 1 for your report, which we will call 
`solution-concentration-effect-on-potato.csv`:

```{r}
table1 <- data.frame(weight = c("initial", "final", 
                                "difference", "percent-change"), 
                     "NaCl-0-percent" = c(NA, NA, NA, NA), 
                     "NaCl-0.9-percent" = c(NA, NA, NA, NA), 
                     "NaCl-2-percent" = c(NA, NA, NA, NA), 
                     "NaCl-5-percent" = c(NA, NA, NA, NA))
write.csv(x = table1, file = "solution-concentration-effect-on-potato.csv", 
          row.names = FALSE)
```

In the above code, make sure to replace the `NA`'s with the values from your experiment. 
Keep in mind that the values entered into the tables should be 
decimals greater than or equal to zero, and likely less than
100 since you measured mass in grams and the potato
cubes for the experiments were 8 cm^3^ in size. 

Notice that we used the file extension ".csv".
CSV stands for comma-separated values, which is a format for 
tabular data stored in a text file where a comma separates the columns of the data. 
That is why a comma separates each of the values in the rows for the different solutions. 

Now, we will make Table 2, which includes the class averages and we will call 
`solution-concentration-effect-on-potato-averages.csv`:

```{r}
table2 <- data.frame(weight = c("percent-change", "minimum-value", 
                                "maximum-value", "difference"), 
                     "NaCl-0-percent" = c(NA, NA, NA, NA), 
                     "NaCl-0.9-percent" = c(NA, NA, NA, NA), 
                     "NaCl-2-percent" = c(NA, NA, NA, NA), 
                     "NaCl-5-percent" = c(NA, NA, NA, NA))
write.csv(x = table1, file = "solution-concentration-effect-on-potato-averages.csv", 
          row.names = FALSE)
```


You might have just done this in Excel. We did not teach this method because...

## Loading or importing data

We will look at two ways of importing data into R, via a command or using the import feature.
Both options are equivalent; and when they are sucsesfull you will see your data as
a new object in the Environment tab (in the Environment/History pane).

### The programmatic way {#programmatic}

We can load our our data files in CSV format into R using the function `read.csv` in the following way:

```{r}
data1 <- read.csv(file = "solution-concentration-effect-on-potato.csv")
data2 <- read.csv(file = "solution-concentration-effect-on-potato-averages.csv")
```

Here, the argument `file =` of `read.csv` is the name of the file we want to read.
Notice that the filename needs to be a character string, so we put it in quotes.


### The RStudio way

In RStudio, click on the Environment tab. 
Then, click on `Import Dataset` and select `From Text (base)`. 

```{r import-feature, fig.cap="Using the Import Dataset feature.", fig.align='center', out.width="80%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/import-feature.png')
```

This feature will open a file browser where you can locate the `.csv` file,
in our case, for Table 1, it is called `solution-concentration-effect-on-potato.csv`.
Once you have selected the file click `Open`. 
A dialog for importing the file will open up:
on the left you can set the options on the import; 
on the top right you can see `Input File`, which is how the file looks like;
on the bottom right you can see `Data Frame`, 
which is a preview of how the file or data will look like in R.
In the `Name` field, write `data1` to rename the dataset.
Set `Heading` to `Yes`, this will 
tell R to read and include the first row of the file as the names of the columns - 
This is an important step! 
You will see that the column names appear in bold in the `Data Frame` box.
When you are ready click `Import`.

```{r import-dialog, fig.cap="Dataset import options.", fig.align='center', out.width="70%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/import-dialog.png')
```

RStudio will now run the R code that imports your dataset. 
It will be similar to the command used in Section \@ref(programmatic).

## Inspecting the data

To take a look at our dataset, we can print it by typing the name in the Console and hitting `Enter` (or `Return`).

```{r}
data1
data2
```

You can also open the dataset by clicking on the object `data1` in the Environment tab. 
This will open a new tab with the dataset, instead of printing it into the Console,
which is a good choice if your dataset is larger.
Another option is to open the dataset using Microsoft Excel from your computer file browser.

<!--chapter:end:06-entering-data.Rmd-->

# Making graphs in R {#graph}

**(AH: Yolanda, I edited a bit here, hope that's okay. Note that ```{r} actually runs the R code and makes the figures, so you don't actually have to import the figures they're generated via R)**
**(YW: I like your changes! Note that there are two references (Murreel 2006 and Abedin and Mittal 2014) in the first section. I don't know how to use the reference command in R, I will put them in manually in the references chapter. 
**(AH: Need instructions to export figures  YW: started that. I'm sure there's room to improve. 
Also, typ = "l" for a line graphs, and also, here I think we should make a note about is x is not monotonically increasing how the plot can get messed up)**

**(AH: Add links to other resources that covers these same topics)

**(AH: Joany do you want to read over this section and meet with me to talk about wht to do to improve it)**

You might have had experience making graphs using `Microsoft Excel`, however, making graphs with `Excel` is not recommended. A number of reasons to choose `R` over `Excel` are given in Chapter \@ref(intro). In addition, the defaults in `Excel` do not meet standards for good scientific graphs and you do not have the option to learn more sophisticated graphs after you master the basics, as the range of options available in `Excel` is limited.

You can make very attractive, clean, professional graphs and figures in `R`. This chapter is a basic introduction to `plot()` functions in `R`. More advanced users can use the package `ggplot`. There are many good reference manuals for making sophisticated graphs (e.g., Murrell 2006; Abedin and Mittal 2014). 

## Scatter and line plots with `plot(x,y)`
In `R`, `plot()` is a built-in function: you do not need to install a package to call it. There are two mandatory arguments `x` and `y` which are vectors that must have the same length. For example,

```{r}
x <- c(2,4,6,8,10)
y <- c(10,20,30,40,50)
```
Copy and paste the above code into your `Console` and type `Enter/Return`. Note that to make a vector we need to use `c()`. Now, query the value of `x` by typing `x` into your `Console`


```{r}
x
```

This lets us know that `x` is a vector that is a list of 5 numbers.

The command `plot(x,y)` will mark points on the graph by taking elements in the same position in the `x` and `y` vectors. For example, in the first position of the `x` and `y` vectors are `2` and `10`, respectively. Because `plot(x,y)` has `x` as its first argument, the values of `x` will correspond to the horizontal axis of the plot. Therefore, the first element of the lists `x` and `y`, will be reflected in the plot as a point marked at the `2` position of the horizontal axis and the `10` position on the vertical axis.

As the vectors `x` and `y` in our example are a list of 5 numbers our graph will consist of 5 points, generated by pairing values in the `x` and `y` vectors in the same position.

``` 
x <- c(2,4,6,8,10)
y <- c(10,20,30,40,50)
plot(x,y)
```

Will give a graph like this:

```{r scatterplot1, fig.cap="A poor example of a scatter plot", fig.align='center', out.width="70%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/Fig_7_1.jpeg')
```

Try it yourself by copy and pasting the above code into your `Console`. If `x` and `y` do not have the same lengths, the `plot()` function will give an error. Try the code below by copy and pasting it into your console:

``` 
x <- c(2)
y <- c(10,20,30,40,50)
plot(x,y)
```
Note the error message that is generated so that if you get this error later when coding on your own you might remember that this error is generated because `x` and `y` have different lengths.

Now try this code:

``` 
x <- c(2,2,2,2,2)
y <- c(10,20,30,40,50)
plot(x,y)
length(x)
length(y)
```
Note we can use `length()` to query the length of `x` and `y`. In addition, the names `x` and `y` are the users choice. A nearly identical plot is generated with the code:

``` 
A <- c(2,2,2,2,2)
B <- c(10,20,30,40,50)
plot(A,B)
```
The key is the internal consistency: `A` and `B` are the names of the vectors and these same names appear as the arguments of `plot(A,B)`. Can you guess what happens if you try `plot(B,A)` or `plot(A,A)`? Type these into your `Console` and give it a try.

---

EXERCISES

1. Which two vectors below will generate a plot if executed before `plot(a,b)`?

```
a = c(1,3,5)
a = c(2,4,5,6,7)
b = c(2,2,4,6)
b = c(2,4,6)
```
2. A friend has written the code below, but it will not run. What do you think the problem could be?

```
a = c(1,2,3)
b = c(2,4,6)
plot(x,y)
```

---

### Adding axes labels and changing symbols
Note that all of the Figures we have made so far are not great: none meet the standards of a good figure. What's missing?

- meaningful `x` and `y` axis labels, 
- maybe it's more appropriate to show this data connected with a line, and
- maybe you want the symbols to be a different shape or colour.

In addition, to the two manadatory arguments `x` and `y`, the `plot()` function has other optional argument that can be called to make a better figure. Try using the `help()` learn about more options to change the look of your plot (refer to Chapter \@ref(help), _Getting help in R_, to recall how to display and read help files):

```
help(plot)
```

You'll see there are a lot of optional arguments. First, let's fix the axis labels.

```{r}
x <- c(2,4,6,8,10)
y <- c(10,20,30,40,50)
plot(x,y, xlab = "temperature (degrees C)", ylab = "growth rate (cm)")
```

This one is a little better. Copy and paste the above code into your `Console` to make sure you can make the graph with axes labels.

If you want to change the symbols, you use the `pch` argument to change the style (see Section \@ref(pch)). Note, that `pch = 17` corresponds to filled triangles, so let's make the graph, but with filled triangles.

```{r}
plot(x,y, pch = 17, xlab = "temperature (degrees C)", ylab = "growth rate (cm)")
```

Note that the order of the optional arguments is unimportant (the function arguments are separated by `,`) particularly where arguments are labelled with `xlab = ` or `ylab = `.  Lets try to see if we can put the arguments out of order and end up with the wrong graph:

```{r}
plot(pch = 17, ylab = "growth rate (cm)", xlab = "temperature (degrees C)", y,x)
```

In the above graph, we wanted `"growth rate (cm)"` and the vertical axis and this has appeared correctly, but the values of growth rate should be `c(10,20,30,40,50)` and now we see the values of `y` have appeared on the horizontal axes rather than on the vertical axis where they should be. This is because `y` appears before `x` in the plot command and whichever vector appears first will be read as corresponding to the values on the horizontal axis. On the other hand `"ylab="` appears before `"xlab="` but this does not cause a problem because `"ylab="` still appears on the vertical axis as it should.

### Line graphs

### Exporting your graphs

Once you've got plots in the "plot" window that you like, you can export them using them so that you can insert them into things like a word document, a power point file or a website. Click on the down error on the "Export" button in the Plots pane in R Studio, then select either to save as an Image or as a PDF. In the "Image" option you can choose various file formats (JPEG, TIFF, PNG, etc.). 

```{r panes, fig.cap="Export button in the Plots Window", fig.align='center', out.width="70%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/ExportPlotWindow.jpg')
```

Be sure to check that you have the correct file path (circled in the figure below) and then give your figure a file name that makes sense (see rectangle in the figure below).

```{r panes, fig.cap="Setting file path (oval - click the Directory button to change the file path) and file name for your exported figure", fig.align='center', out.width="70%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/ExportPlot.jpg')
```

## Other usesful plot functions

The generic `plot()` function makes a scatter plot, with variations. 

For other types of plots, R has different functions, but all have similar arguments for adjusting the labels, spacing, tic marks, symbols, etc. 

See the following

```barplot{graphics}``` to make bar graphs
```boxplot{graphics}``` to make boxplots
```hist{graphics}``` to make a histogram


## Getting a bit more sophisticated

It's not hard to get more sophisticate with R graphics. The best approach is to explore one of the reference books listed to figure out how to do what you want to do.

One thing most undergrads will probably need or want to do is to visualize two sets of data on the same graph. For example, the goldfish data in the Appendix of the BIOL 1001 lab manual shows the opercular beats of two fish in response to temperature. It would be logical to show both fish on the same graph, but with different symbols. Here's how you would do it:

```{r} temp <-cbind(5, 8, 12, 16, 20, 23)```
This creats the vector of data for the x-axis.


```{r} temp <- cbind(5, 8, 12, 16, 20, 23)```

This creates the vector of data for the x-axis

```{r} fishA <- cbind(36, 42, 59, 73, 88, 96)```
This creates the vector of data for Fish A```

```{r} fishB <- cbind(40, 49, 61, 71, 82, 90)``` 
This creates the vector of data for Fish B```

First we plot x and y with the "temp" vector as "x" and the "fishA" vector as y 

```{r} plot(temp, fishA, pch = 17, xlab = "temperature (degrees C)", ylab = "opercular beats (bpm)")```

Then to add a second vector of data (fishB) to the same plot, we use:
```{r} points(temp, fishB, pch = 17, col = "dimgray")```


Voila!

```{r goldfishplot, fig.cap="The effect of water temperature on opercular beats per minute (bpm) for two goldfish; goldfish A (black triangles) and goldfish B (grey triangles)", fig.align='center', out.width="70%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/Fig_7_4.jpg')
```

Note that the difference in the two symbols is described in the caption. For most scientific journals it is preferred to describe the symbology in the caption, rather than use a legend. However, with more complex graphs, or for some publications, you may wish to include a legend. Legends can be a bit tricky to work with in R - we'll leave those for a more advanced course. 

Finally, if you wanted to have a line joining the points, you'd specify the ```type = "l"``` in the arguments of the plot fuction (l is for line, p is for points, b is for both).

```{r} plot(temp, fishA, pch = 17, type = "l", xlab = "temperature (degrees C)", ylab = "opercular beats (bpm)")```

## pch cheatsheet {#pch}
As it is hard to remember that `pch = 17` corresponds to filled triangles you can make a little cheat sheet for `pch` codes using the code below. You may want to print out a chart of the `pch` codes and tape it up on your wall by your computer, or in the front of your notebook as a handy reference.

```{r}
#++++++++++++++++++++++++++++++++++++++++++++
#generate a plot of point shapes which R knows about.
#++++++++++++++++++++++++++++++++++++++++++++
generateRPointShapes<-function(){
  oldPar<-par()
  par(font=2, mar=c(0.5,0,0,0))
  y=rev(c(rep(1,6),rep(2,5), rep(3,5), rep(4,5), rep(5,5)))
  x=c(rep(1:5,5),6)
  plot(x, y, pch = 0:25, cex=1.5, ylim=c(1,5.5), xlim=c(1,6.5), 
       axes=FALSE, xlab="", ylab="", bg="blue")
  text(x, y, labels=0:25, pos=3)
  par(mar=oldPar$mar,font=oldPar$font )
}
generateRPointShapes()
```

<!--chapter:end:07-graph-in-R.Rmd-->

# Hypothesis, predictions and equations {#hypothesis}

The proposed answer to a question is termed a **hypothesis**. Hypotheses are formulated on the basis of some preliminary observations and/or prior knowledge.

How does the presence of a concentration gradient affect the diffusion of water? There are three possible answers to this question. Indeed, many questions may generate multiple alternative hypotheses.

(**we are considering changing these examples to be about moose to make them more fun**)
<div class="alert alert-info">
<strong>Hypotheses</strong>

<strong>H1</strong>.	Water diffuses from a hypertonic solution into a hypotonic solution.
	
<strong>H2</strong>.	Water diffuses from a hypotonic solution into a hypertonic solution.
	
<strong>H3</strong>.	Water diffuses equally in both directions along the osmotic gradient.
</div>

All three meet the basic requirement of a hypothesis, namely, they are predictive, are clearly expressed and are an answer to the problem; however, #2 is the most likely hypothesis given background reading on osmosis.

**Predictions** are generated from the hypothesis and framed in terms of a specific experiment or analysis. Let's define a specific experiment:
	
---	

EXPERIMENT 1

An osmometer containing a 30% glucose and dye solution is placed in a beaker containing distilled water.

---

If hypothesis #2 holds, then true water will diffuse from the hypotonic solution (the water in the beaker) into the hypertonic solution (the 30% glucose and dye solution in the osmometer) and the water will rise in the pipette. Note how this prediction was generated using the if ... then ... format.

i.e.	*if* the hypothesis holds true, *then* a specific event will occur (prediction).

<div class="alert alert-info">
The <strong>prediction</strong> corresponding to H2 for Experiment 1 is:  <br>

*if* water diffuses from a hypotonic solution into a hypertonic solution (hypothesis), <br>
*then* water will diffuse from the beaker into the osmometer causing the water in the pipette to rise.
</div>
	
Although  hypotheses and predictions are similar, two basic differences exist: hypotheses are **general**, predictions are **specific**; hypotheses are written in the present tense, predictions in the future tense.
	
**Equations** write the predictions in mathematical notation so that statistical tests can be applied to determine if the hypothesis is supported. The equation corresponding to the prediction in the box above is:

<div class="alert alert-info">
The <strong>equation</strong> corresponding to the prediction above is:

$v_F > v_0$ <br>

where $v_F$ is the final volume of the glucose-dye solution in the osmometer (units: mL), and <br>
$v_0$ is the initial volume of glucose-dye solution in the osmometer (units: mL).
</div>

*	When writing your prediction as an equation *you* get to choose the notation: you might have instead decided that *y* is the final volume of the glucose-dye solution; this choice is also correct if you correctly explain what variable is represented by what you have chosen to call *y*. You many choose any lower or upper case Arabic (i.e., *a,b,c*‚Ä¶) or greek (i.e., Œ±, Œ≤, Œ≥, ‚Ä¶) letters for your variables and you may also choose to use subscripts.

*	Choose symbols that are one letter long. For example, $vF$ (two symbols long) is a less preferred symbol choice because this could be confused with $v$ multiplied by $F$.

*	It is usual to choose symbols that are intuitive (i.e. here, $v$ helps us think of volume and $F$ helps us think of ‚Äòfinal‚Äô). Commonly, Œº or bar notation, i.e., $\bar{\nu}$ represents the mean of multiple observations. It is a good idea to avoid ‚ÄúO‚Äù which can be confused with zero, and e, i, œÄ and Œ£, which have specific definitions in mathematics, however, these considerations are preferred, an answer that does not use these conventions is not necessarily wrong. The necessary element of your answer is that you provide a description of the notation you define.

*	You must provide the units for the quantities you define.

*	The dependent variable (see below for a defintion) is to appear on the lefthand side of the equals or inequality.

You will not perform statistical tests in BIOL 1001 or 1002. However, learning how to write your predictions as an equation will help you to understand how your hypothesis and predictions connect to statistical tests that help determine whether a hypothesis is supported or refuted.
	
##	Discrete and continuous variables
	
Generally, a hypothesis describes the effect of one factor on another, for example, the effect of a concentration gradient on the direction of osmosis. These factors are called **variables** and are of two types: the variable that is changed (varied) by the experimenter is called the **independent variable**, and the variable that changes as a consequence of the changes to the independent variable is called the **dependent variable**. In our example:
	
*	The glucose concentration in the osmometer is the **independent variable** because it is created by the experimenter, and

*	The volume of water in the pipette is the **dependent variable** because it is affected directly by the concentrations inside and outside the osmometer, and is the measured response.

---

EXPERIMENT 2


This experiment will test how concentrations of 30, 60, and 90% glucose in an osmometer will influence the flow of water molecules across two solutions. 

---
	
For Experiment 2, we might have the following as our hypothesis, prediction and equation:

<div class="alert alert-info">
<strong>Hypothesis:</strong> When a solution is more hypertonic, more water will diffuse into it from a hypotonic solution.
	
<strong>Prediction:</strong> The volume of water in the osmometer will be largest for the 90% glucose solution, and smallest for the 30% glucose solution.
	
<strong>Equation:</strong>

$v_F = bg + v_0$,  

where $v_F$ is the final volume of liquid in the osmometer (mL), $b$ is the change in final volume per unit glucose concentration (mL per %), g is the glucose concentration (%), and $v_0$ is the final volume when the glucose concentration is zero (mL).

</div>

Note that in Experiment 2, the independent variable, glucose concentration, takes three different values, (30, 60 and 90%) and these values are taken from options along a continuum. As such, in the above equation we have treated glucose concentration as a CONTINUOUS variable. This changes our hypothesis and predictions. Let‚Äôs examine the new equation more closely.
	
*	We predicted an increasing relationship between glucose concentration and the final volume of water in the osmometer. In writing our equation, we, more specifically, assumed a linear, or ‚Äústraight-line‚Äù, relationship. An increasing relationship does not have to be a linear relationship, however, without a good reason to assume otherwise, we should assume that a prediction stating an increasing or decreasing relationship between two variables corresponds to a linear (or ‚Äústraight-line‚Äù) relationship with where an increasing relationship has a positive slope and a decreasing relationship has a negative slope.
  
* Recall that the equation for a straight-line is: $y = m x + b$,
where $y$ corresponds to the vertical axis of a graph, $x$ corresponds to the horizontal axis, $m$ is the slope of the line and $b$ is the y-axis intercept (corresponding to $x = 0$). Note that the *equation* above is simply the equation for a straight-line written in the notation of our experiment:
    + $v_F$, the dependent variable, is $y$,
    + $b$, the effect of $g$ on $v_F$, is $m$, the slope,
    + $g$, the independent variable, is $x$,
    + $v_0$, the final volume when $g = 0$, is $b$.
  
*	Note that the units for each term are equal:
    + $v_F$ has units mL
    + $bg$ has units $\frac{mL}{\%} \cdot \% =$ mL
    + $v_0$ has units mL
	Finally, lets consider a third experiment where the independent variable is *discrete*. 
	
---

EXPERIMENT 3

The Glycemic Index Foundation of South Africa designates Beverages as ‚ÄúLow‚Äù, ‚ÄúIntermediate‚Äù or ‚ÄúHigh‚Äù Glycemic Index (GI) values. In this experiment, 3 beverages from each group: Low, Intermediate, and High, are used in place of the glucose solution in the osmometer, and for each of these 9 experiments the solution in the beaker is distilled water.

---
	
In Experiment 3, the independent variable is **discrete** because each beverage belongs to a group. The hypothesis, prediction, and equation for Experiment 3 might be:

<div class="alert alert-info">
<strong>Hypothesis:</strong> When a solution is more hypertonic, more water will diffuse into it from a hypotonic solution.  

<strong>Prediction:</strong> The volume of water in the osmometer will be largest for the three beverages from the ‚ÄúHigh‚Äù GI group, and smallest for the 3 beverages from the ‚ÄúLow‚Äù GI group.  

<strong>Equation:</strong>

$Œº_H > Œº_I > Œº_L$,

where $Œº_H$ is the mean volume of liquid in the osmometer for the three ‚ÄúHigh‚Äù GI beverages (mL), $Œº_I$ is the mean volume of liquid in the osmometer for the three ‚ÄúIntermediate‚Äù GI beverages (mL), and $Œº_L$ is the mean volume of liquid in the osmometer for the three ‚ÄúLow‚Äù GI beverages (mL).

</div>

*	The equation for Experiment 3 is similar to the equation for Experiment 1. For Experiment 3, there were three beverages in each group, so we defined a measure of the central tendency, in this case the mean, so that we could compare between the groups.  

## Hypothesis testing

### Null hypotheses
Null means nothing or zero and null hypotheses usually describe outcomes where the independent variables have no effect. Null hypotheses are usually paired with alternative hypotheses whereby the independent variable does have an effect. As such, H3 is a potential null hypothesis, since this hypothesis describes the concentration gradient having no effect on the direction of difussion, and could be paired with the alternative hypothesis, H2. When written as an equation, frequently null hypotheses are expressed as a given parameter set equal to zero.

<div class="alert alert-info">

<strong>Null Hypothesis:</strong> Water diffuses equally in both directions along the osmotic gradient.

<strong>Alternative Hypothesis</strong>.	Water diffuses from a hypotonic solution into a hypertonic solution.
	
<strong>Prediction under the null hypothesiss:</strong> The volume of water in the osmometer will not change.
	
<strong>Equation under the null hypothesis:</strong>

$v_F = v_0$,  

where $v_F$ is the final volume of the glucose-dye solution in the osmometer (units: mL), and <br>
$v_0$ is the initial volume of glucose-dye solution in the osmometer (units: mL).
</div>

Historically, many statistical analyses have been performed in the null hypothesis testing framework. More recently, information theoretic approaches that assess the relative support for each of multiple working hypotheses have gained popularity (Burnham and Anderson 2003); that approach may be introduced in upper year courses. 

## Further reading

<!--chapter:end:09-writing-equations.Rmd-->

# BIOL 1001: LABORATORY 2 {#lab2}

## BIOLOGICAL DATA WITH COMPUTERS {-}

---

PURPOSE

1. To familiarize with R Studio
1. To learn how to record data in electronic format
1. To learn how to write hypotheses as equations
1. To learn how to choose the appropriate visualizations
1. To learn how to make graphs using R Studio

---



Before coming to the laboratory:

1. If you wish to use your own laptop, then install `R` and `RStudio`. For instructions see Chapter \@ref(install).
 
1. Before coming to the laboratory read:
 
In the quant guide: 

- Sections

In the printed lab manual:

- Appendix II: Something or other
 
## EXERCISE 1.	Intro to RStudio {-}
Maybe borrow a little exercise from software carpentary. Nothing to hand in/assess?

## EXERCISE 2.	Entering data {-}

We will use data from a research paper that investigated 
whether plants were able to respond to stress cues emitted from their drought-stressed neighbours
[@falik2011rumor].
The research team used _Pisum sativum_ plants in an experimental setup that allowed them to connect the plants' root systems. One plant in the central position of a row was subjected to osmotic stress, while neighbouring plants remained unstressed on both sides. 
Plants on one side of the stressed plant shared roots with others in the group but
did not share roots with the stressed plant; 
these were the control group. 
On the other side, the stressed plant shared its roots with its nearest unstressed neighbour, and all the other plants shared their roots with their nearest neighbour
(See Figure \@ref(fig:setup)).

```{r setup, fig.cap="Experimental set up for testing stress cues. Circles represent plants and connector lines represent the plants with connected roots. The numbers of the plants correspond to those of Table 9.1. Osmotic stress was induced in plant 6 (Modified from Falik et al., 2011).", fig.align='center', out.width="70%", fig.show='hold', echo=FALSE}
knitr::include_graphics('./figures/Falik_et-al_2011_setup.png')
```


Stress was quantified fifteen minutes after the induction of drought by measuring the width of the stomatal openings on the leaves of the plants. The results of this experiment are as follows:

```{r raw-data, echo = FALSE, results = 'asis'}
library(knitr)
dataset <- read.csv("Falik_et-al_2011_dataset.csv")
kable(dataset[1:3], caption = "Raw data from Falik and collaborators (2011).")
```

### Questions {-}
1. Follow the instructions in Chapter \@ref(data-entry) and enter the data into `R` and then save the data as a .csv. Submit your .csv file to the `Biology Data Repository` Brightspace.

1. Follow the instructions in Chapter \@ref(data) and write metadata for the data that you saved in .csv format. Submit your metadata to the `Biology Data Repository` Brightspace.

1. Looking at the dataset, can you tell how many treatments were applied in the experiment? How many plants were used for each treatment?

1. What is the smallest stomatal width measured? What is the largest?

## EXERCISE 3.	Discrete independent variable {-}
### Questions {-}

1. Follow the instructions in Chapter \@ref(graph). Replicate the figure from the Campbell textbook (bar plot). Export your bar graph as a .pdf and insert it into Microsoft Word (or similar software). Add a caption to your bar plot. Print your graph to hand in with your lab report.

1. In the bar graph, what can you tell about the stomatal openings of the treatment plant vs. the control plants? Is it consistent across all the individual plants?

1. Follow the instructions in Chapter \@ref(graph) to re-plot the data from EXERCISE 1 as a boxplot that compares control versus treatment. Export your boxplot as a pdf file and insert it into Microsoft Word (or similar software). Add a caption to your boxplot.

1.	Label the following on the boxplot: mean, 25% and 75% quantiles, 95% confidence limits, outliers. It is acceptible to write these labels in with pen or pencil.

1.	What does the boxplot tell you about the differences between the treatment plants (plants 6-11) vs. the control plants that the treatment plant does not?

1.	Why do you think a scientist might choose to represent these data one way over another?


### Write the hypothesis as an equation {-}
### Questions {-}
1. Follow the instructions in Chapter \@ref(hypothesis). Write a hypothesis, prediction, and equation for these data.


## EXERCISE 4. Continuous independent variable {-}

The increase of carbon dioxide in the atmosphere is expected to reduce the concentrations of carbonate ions in the surface of the ocean. This process, in general called ocean acidification, affects the biology and survival of a wide range of marine species because it prevents the formation of shells and plates of calcium carbonate.
A long term study on coral reefs showed how the net community calcification rate (G, mmol CaCO~3~m^-2^d^-1^) is related to the concentration of carbonate (CO~3~^2-^, $\mu$mol$\cdot$kg^-1^) [@langdon2000effect]. 
We will use the data from the research paper to understand this relationship.
The results of this experiment are as follows:

```{r langdon-data, echo = FALSE, results = 'asis'}
library(knitr)
dataset <- read.csv("Langdon_et-al_2000_dataset.csv")
kable(dataset[1:11,1:2], caption = "Raw data from Langdon and collaborators (2000).")
```

### Make a graph {-}

1. We provide this data as a .csv called `Langdon_et-al_2000_dataset.csv`. Follow the instructions in Chapter \@ref(data-entry) and import the data.

1. Follow the instructions in Chapter \@ref(graph) to make a scatterplot for these data and replicate the figure from the Campbell textbook.

1. Follow the instructions in Chapter \@ref(graph) to make a line graph for these data. Export your line graph as a pdf file and insert it into Microsoft Word (or similar software). Write a figure caption. Print your graph and hand it in with your lab report for this week.

1. Follow the instructions in Chapter \@ref(hypothesis). Write a hypothesis, prediction, and equation for these data.


<!--chapter:end:11-BIOL-1001-lab-2.Rmd-->

# Coding club at MUN

Starting in Fall 2020, we will begin a weekly coding club aimed at undergraduates in Biology (or related disciplines) that would like to improve their coding skills.

During the coding club you might like to:

- Re-read this manual and make sure you are able to complete the exercises.

- Work on software carpentary modules: https://software-carpentry.org/lessons/

- Work on coding club modules: https://ourcodingclub.github.io/

- Work on University of Toronto coders modules: https://uoftcoders.github.io/studyGroup/lessons/

- Work on the Tidy Tuesday problem: https://github.com/rfordatascience/tidytuesday

- ADVANCED: Help us with the development of this manual. Specifically, if you can learn how to use `Github` and `bookdown` you can easily incorporate your suggestions. The repository for this training manual is here: https://github.com/ahurford/quantitative-training-guide

- ADVANCED: Help others with `R`. Gradaute students are also welcome to attend coding club.

<!--chapter:end:19-coding-club.Rmd-->

`r if (knitr:::is_html_output()) '
# References {-}
'`


YW to JM: There are some references in the document not linked here... Burnham and Anderson 2003 (in chapter 9) and two references in chapter 8. I've put them in manually below, maybe at some point you can show me how to work the referencing in bookdown... thanks!

Abedin, Jaynal and Hrishi Mittal. 2011. R graphics cookbook. 2nd edition. Packt Publishing.

Burnham, Kenneth P., and David R. Anderson. 2003. Model selection and multimodel inference: a practical information-theoretic approach. Springer, NY. 

Murrell, Paul. 2006. R graphics. Chapman and Hall, CRC Press. Available [here](https://www.stat.auckland.ac.nz/~paul/RG2e/)


<!--chapter:end:20-references.Rmd-->

